File: python/trees/kth-smallest-element-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        array = []

        def inorder(node):
            if not node:
                return

            left = inorder(node.left)
            if left:
                array.append(left.val)
            array.append(node.val)
            right = inorder(node.right)
            if right:
                array.append(right.val)

            return

        inorder(root)
        return array[k - 1]

File: c++/trees/kth-smallest-element-in-a-bst.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        getValues(root);
        return elements_[k-1];
    }

private:
    void getValues(TreeNode* node) {
        if (!node) {
            return;
        }

        getValues(node->left);
        elements_.push_back(node->val);
        getValues(node->right);
    }

    std::vector<int> elements_;
};
